lib/recursive-methods.rb

```diff
@@ -1,49 +1,112 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n is the length
+# Space complexity: O(n), uses up some space in the call stack until it reaches the base case
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    if n == 0
+        return 1
+    elsif n <= 0
+        raise ArgumentError
+    else
+    return n * factorial(n-1)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2)
+# Space complexity: O(n^2)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    if s.length <= 1
+        return s
+    end
+
+    return s[-1] + reverse(s[1..-2]) + s[0]
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+    first = 0
+    last = s.length - 1
+    return reverse_inplace_helper(s, first, last)
+end
+
+def reverse_inplace_helper(s, first = 0, last = s.length - 1)
+    if first > last
+        return s
+    end
+
+    temp = s[first]
+    s[first] = s[last]
+    s[last] = temp
+
+    return reverse_inplace_helper(s, first + 1, last - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+    if n == 0
+        return 0
+    end
+
+    return 2 + bunny(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2)
+# Space complexity: O(n^2)
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+    if s.length < 1
+        return true
+    elsif  s == "(" || s == ")"
+        return false
+    elsif s[0] == "(" && s[-1] == ")"
+        return nested(s[1..-2])
+    else
+        return false
+    end
+
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2)
+# Space complexity: O(n^2)
 def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+    if array.empty?
+        return false
+    elsif value == array[0]
+        return true
+    else
+        return search(array[1..-1], value)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2)
+# Space complexity: O(n^2)
 def is_palindrome(s)
```

Comment:
👍 However can you see a way, like with reverse_inplace to do it with O(n) time/space complexity?